BF1942 and BFV process detection and con file dynamic file path lookup

AnomalousNicole opened the pull request:
**Recently, I installed Battlefield Rich Presence and noticed that it did not show game info in Discord while playing BF1942 and BFV.  I cloned the repos for Battlefield Rich Presence and GameDataReader and debugged the code to figure out what was happening.**


**BF1942 had the following Battlefield Rich Presence issue that prevented game information from displaying in my profile in Discord.**

**Issue:** Battlefield Rich Presence was not able to find the BF1942 game process using a RegEx to find the running process.

**Fix:** I modified Battlefield Rich Presence to look for any processes named "BF1942" and if one match is found then a GameInfo object will be returned from the IsRunning method.

**Testing:** After the fix above was applied to Battlefield Rich Presence, I built a debug version of Battlefield Rich Presence and ran BF1942 again.  This time when BF1942 was run, Battlefield Rich Presence was displayed in my Discord profile.  However, it only displayed the "In the menus" status in my profile.  When I debugged the Battlefield Rich Presence further, I was able to determine that GameDataReader was passing back the message "Couldn't find configuration data. Is the game installed?" to Battlefield Rich Presence.  This is what led me to debug the GameDataReader code.


**BF1942 and BFV each had the GameDataReader issue below that was causing the con files for each game to not be found correctly.**

**Issue:** GameDataReader was not able to find the GeneralOptions.con and Profile.con files located within the "mods\bf1942\settings" and "Mods\BfVietnam\settings" folders in each game install folder.  GameDataReader was looking in the "%LOCALAPPDATA%\VirtualStore\Program Files (x86)\EA Games\Battlefield 1942\Mods\bf1942\settings" and "%LOCALAPPDATA%\VirtualStore\Program Files (x86)\EA Games\Battlefield Vietnam\Mods\BfVietnam\settings" folders for each game.  These folders are used when BF1942 or BFV are installed under Program Files (x86) and the game does not have permission to write to the game install folder.  However, when BF1942 or BFV is not installed under Program Files (x86) each game uses the "mods\bf1942\settings" or "Mods\BfVietnam\settings" located in that game's install folder.

**Fix:** I modified GameDataReader to dynamically lookup the path where the BF1942 or BFV processes are running from.
Those paths are then used to check if the GeneralOptions.con and Profile.con files exist within the "mods\bf1942\settings" and "Mods\BfVietnam\settings" folders in each game install folder.
If the GeneralOptions.con and Profile.con files are not found, then GameDataReader will check for the existence of the con files in the "%LOCALAPPDATA%\VirtualStore\Program Files (x86)\EA Games\" for each game.

**Testing:** After the fix above was applied to GameDataReader, I built a debug version of GameDataReader and then copied the GameDataReader DLL to the debug folder for Battlefield Rich Presence.  I then ran Battlefield Rich Presence and also launched BF1942 and connected to a multiplayer server.  Battlefield Rich Presence then correctly displayed the game and server information in my Discord profile.  I repeated the same test for BFV and Battlefield Rich Presence also displayed the game and server information in Discord.


**Note:** The fix I created for Battlefield Rich Presence is dependent upon the fix for GameDataReader.
In each program I have created my fixes in a branch called **"bf1492-bfv-fixes"**.
I have submitted pull requests on both the Battlefield Rich Presence and GameDataReader GitHub repo pages.


Below are screenshots of Battlefield Rich Presence displaying game information in Discord for Battlefield 1942 and Battlefield Vietnam.

**Battlefield 1942**
![BattlefieldRichPresence-BF1942](https://user-images.githubusercontent.com/481352/221685296-d3b1b08e-e05e-4aea-9320-9016c99bd417.jpg)

**Battlefield Vietnam**
![BattlefieldRichPresence-BFV](https://user-images.githubusercontent.com/481352/221685298-7933496f-f885-42d3-a50d-211d9e24b085.jpg)


**Hopefully, the above information has been helpful in explaining the issues that were fixed in each program.  Please let me know if more information is needed.**

AnomalousNicole committed:
fix: Get the bf1942 and bfv Profile.con and GeneralOptions.con file locations dynamically based on where the game process is being executed from.  If the game process cannot be found, then fall back to the AppData folder paths.

AnomalousNicole committed:
fix: Added logic to check for the existence of the bf1942 and bfv Profile.con and GeneralOptions.con files in the executing and fallback file paths

cetteup commented on src/GameDataReader/Common/Refractor/V1/Files/GlobalRefractorV1ConfigFile.cs:
```diff
@@ -5,17 +5,22 @@ namespace GameDataReader.Common.Refractor.V1.Files;
     {
+        var utilityMethodsRefractorV1 = new UtilityMethodsRefractorV1ConfigFile();
         _modName = modName;
+        _gameProcessPath = utilityMethodsRefractorV1.GetProcessPath(_modName);
+        var conFile = $@"\Mods\{_modName}\Settings\Profile.con";
+        _conFilePath = utilityMethodsRefractorV1.GetConFilePath(conFile, _gameProcessPath, GameName);
```
Since this is building the path to the config file, this logic would be better placed in `GetFilePath`. I don't see any immediate benefit of having it in the constructor.

cetteup commented on src/GameDataReader/Common/Refractor/V1/Files/UtilityMethodsRefractorV1ConfigFile.cs:
```diff
@@ -0,0 +1,35 @@
+
+namespace GameDataReader.Common.Refractor.V1.Files;
+
+internal class UtilityMethodsRefractorV1ConfigFile
+{
+    public string GetProcessPath(string modName)
+    {
+        var matchingProcesses = Process.GetProcessesByName(modName);
+        return matchingProcesses.Length == 1 ? matchingProcesses[0].GetMainModuleFileName(modName) : string.Empty;
+    }
+
+    public string GetConFilePath(string conFile, string gameProcessPath, string gameName)
+    {
+        var appData = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
+        var executingPath = File.Exists($@"{gameProcessPath}{conFile}") ? $@"{gameProcessPath}{conFile}" : string.Empty;
+        var fallbackPath = File.Exists($@"{appData}\VirtualStore\Program Files (x86)\EA GAMES\{gameName}{conFile}") ? $@"{appData}\VirtualStore\Program Files (x86)\EA GAMES\{gameName}{conFile}" : string.Empty;
+        return File.Exists($@"{gameProcessPath}{conFile}") ? executingPath : fallbackPath;
+    }
+}
```
Entire class could be static, unless I am missing something. Would eliminate the need to instantiate any instances of the class.

TwitchPlaysBF2 replied:
Keep it non-static please, like the other config files.

It's preferred to avoid static classes, unless the class is responsible for only holding data, or doing minor stuff, which won't get extended, tested or replaced (as a dependency to something else)

Reason being, if someone wants to test the logic around it, you cannot inject outside dependencies (through constructor) or replace the class as a dependency itself to another unit of code, with a mocked implementation of it (cannot make an interface on static class)

cetteup replied:
This class isn't a config file, it's just holding some util methods. Maybe the name is a bit misleading?

TwitchPlaysBF2 replied:
I missed that, sorry, you're right

Then let's rename the file to `RefractorV1PathResolver` or something like that

cetteup commented on src/GameDataReader/Common/Refractor/V1/Files/UtilityMethodsRefractorV1ConfigFile.cs:
```diff
@@ -0,0 +1,35 @@
+
+internal class UtilityMethodsRefractorV1ConfigFile
+{
+    public string GetProcessPath(string modName)
```
While the attribute of the config file is called `_modName`, I don't think it makes sense to use that as parameter name here. The method is totally generic, so it should be something like `processName`.

cetteup commented on src/GameDataReader/Common/Refractor/V1/Files/UtilityMethodsRefractorV1ConfigFile.cs:
```diff
@@ -0,0 +1,35 @@
+    public string GetProcessPath(string modName)
+    {
+        var matchingProcesses = Process.GetProcessesByName(modName);
+        return matchingProcesses.Length == 1 ? matchingProcesses[0].GetMainModuleFileName(modName) : string.Empty;
```
I know @TwitchPlaysBF2 is not a fan of ternary operators, since they are a bit harder to read than regular if statements. So I'd refactor this to a regular if statement.

Also, it is possible to launch multiple instances of Refractor games. While that is probably a total edge case, it can be handled by simply checking whether the process list contains any items (instead of checking it contains exactly one item).

Both put together might look something like this:

````c#
if (matchingProcesses.Length > 0) {
    return matchingProcesses[0].GetMainModuleFileName(modName);
}
return string.Empty;

TwitchPlaysBF2 replied:
Before this gets touched I am wondering what this DLL function call is for. Does it have to do with the parent process or something? 

If not, and the goal is, to just read the process directory, then this would just do it, without that dll import or the `GetMainModuleFileName` method:

```csharp
public string GetProcessPath(string modName)
{
    var foundProcess = Process.GetProcessesByName(modName).FirstOrDefault();
    var filePath = foundProcess?.MainModule?.FileName;
    var folderPath = Path.GetDirectoryName(filePath);

    return folderPath ?? string.Empty;
}
```

cetteup commented on src/GameDataReader/Common/Refractor/V1/Files/UtilityMethodsRefractorV1ConfigFile.cs:
```diff
@@ -0,0 +1,35 @@
+    public string GetConFilePath(string conFile, string gameProcessPath, string gameName)
+    {
+        var appData = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
+        var executingPath = File.Exists($@"{gameProcessPath}{conFile}") ? $@"{gameProcessPath}{conFile}" : string.Empty;
+        var fallbackPath = File.Exists($@"{appData}\VirtualStore\Program Files (x86)\EA GAMES\{gameName}{conFile}") ? $@"{appData}\VirtualStore\Program Files (x86)\EA GAMES\{gameName}{conFile}" : string.Empty;
+        return File.Exists($@"{gameProcessPath}{conFile}") ? executingPath : fallbackPath;
```
This seems a bit chaotic. You are building all paths multiple times and are even checking twice if `$@"{gameProcessPath}{conFile}"` exists. This could much cleaner with regular if statements and early returns, e.g.

```c#
var executingPath = $@"{gameProcessPath}{conFile}";
if (File.Exists(executingPath)) {
    return executingPath;
}
...
return string.Empty;
```

That was you get rid of the ternary operators, only have build each path once and only check if the path exists. Plus you don't do any additional I/O operations once you already know you'd be returning `executingPath`.

And on a more general note: I am a big proponent of using path libraries to join paths (or manipulate paths in any other way). You'll usually end up with much more robust code than when simply concatenating strings. I know GameDataReader currently simply uses string concatenation everywhere, so that is something to maybe change in a later refactor.

cetteup commented on src/GameDataReader/Common/Refractor/V1/Files/UtilityMethodsRefractorV1ConfigFile.cs:
```diff
@@ -0,0 +1,35 @@
+    {
+        var fileNameBuilder = new StringBuilder(buffer);
+        var bufferLength = (uint)fileNameBuilder.Capacity + 1;
+        return (QueryFullProcessImageName(process.Handle, 0, fileNameBuilder, ref bufferLength) ? Path.GetDirectoryName(fileNameBuilder.ToString()) : string.Empty) ?? string.Empty;
```
Harder to read than it needs to be due to a) the ternary operator and b) the fallback value at the end.

TwitchPlaysBF2 replied:
I personally only find ternaries ugly, if they making lines too long, or have too many nested operators.

And since it's always nicer for debugging to have simple return values, I'd do something like this:

```csharp
var filePath = QueryFullProcessImageName(process.Handle, 0, fileNameBuilder, ref bufferLength) 
    ? Path.GetDirectoryName(fileNameBuilder.ToString()) 
    : string.Empty;

return filePath ?? string.Empty;
```

cetteup commented:
Nice work! Very detailed description of the underlying issue. 👍 

I added a few comments regarding some details. I do, however, see one bigger issue with the current approach: **It only works if the game is running.** That's fine when GameDataReader is used by something such as BattlefieldRichPresence, but it is neither obvious to any other consumer nor is it the case for any other game supported by this library.

Since the running game process is only used to determine the path the game is running from, a better approach could be to simply get the install path from the registry. This has the drawback of only working if the game is installed properly, but it does ofc work regardless of whether the game is running. The usual registry path when installed in x86 would be

```
HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\EA Games\Battlefield 1942
```
That path might look different for you with the game installed elsewhere, which is something to check. Whatever the path, it contains a key called `GAMEDIR` which points to the install directory. That path could be used instead of the path of any running process.

AnomalousNicole commented:
@cetteup Thank you for the feedback on my changes.  I am working on implementing the changes you have outlined in your comments.  

While testing the latest changes to Battlefield Rich Presence, I did discover another bug with GameDataReader in relation to the playerName that gets passed back to Battlefield Rich Presence. BF1942 and BFV use ANSI character encoding for the CON files. In one of my player names, I use the registered trademark symbol "®". When GameDataReader passes back the player name with the "®" symbol in it, the URL that gets built by Battlefield Rich Presence does not work correctly. I was able to fix that issue in GameDataReader by using the System.Text.Encoding.CodePages NuGet package and making the change below in the ReadConfigFile method in LineBasedConfigFile.


**//BF1942 and BFVietnam CON files use ANSI/windows-1252 character encoding
var configLines = new List<string>();
if (filePath.ToLower().Contains("bf1942") || filePath.ToLower().Contains("bfvietnam"))
{
    Encoding.RegisterProvider(CodePagesEncodingProvider.Instance);
    configLines = File.ReadAllLines(filePath, Encoding.GetEncoding("windows-1252")).ToList();
}
else
{
    configLines = File.ReadAllLines(filePath).ToList();
}**


BF1942/BFV Player Name In CON File: lookoutriddick®
Player Name Passed Back From GameDataReader: lookoutriddick�

Non working URLs:
https://api.bflist.io/bf1942/v1/players/lookoutriddick%EF%BF%BD/server
https://api.bflist.io/bfvietnam/v1/players/lookoutriddick%EF%BF%BD/server

Working URLs:
https://api.bflist.io/bf1942/v1/players/lookoutriddick%C2%AE/server
https://api.bflist.io/bfvietnam/v1/players/lookoutriddick%C2%AE/server

AnomalousNicole committed:
fix: Read the bf1942 and bfv CON files using windows-1252 character encoding

AnomalousNicole commented:
@cetteup I have pushed a commit the resolves the character encoding issue when the bf1942 and bfv con files are read by GameDataReader.  I am working on fixes for the other items noted in your comments from my original commits.  I plan to push those fixes sometime tomorrow.

TwitchPlaysBF2 commented on src/GameDataReader/Common/Refractor/V1/Files/UtilityMethodsRefractorV1ConfigFile.cs:
```diff
@@ -0,0 +1,35 @@
+    private static extern bool QueryFullProcessImageName([In] IntPtr hProcess, [In] uint dwFlags, [Out] StringBuilder lpExeName, [In, Out] ref uint lpdwSize);
+
+    public static string GetMainModuleFileName(this Process process, string modName, int buffer = 1024)
+    {
```
`modName` param is not being used

TwitchPlaysBF2 reviewed (commented):
Hello @johnbrechbuhl85  

I am very pleasantly surprised at the quality of your contribution summary, and the effort you went to fix the underlying issues. 

I added some of my opinions as a review, if you want to check them out. But considered that I am in a clean-code/over-engineering phase of my career, take it with a grain of perfectionism-salt. 

Feel free write non-perfect code, we're not building large enterprise scale software here. My only true passion is, to make it as readable as possible. Always nicer to debug/understand code this way.

Thanks at this point to @cetteup for the review and your past contributions, you are actually my driving force behind my initiatives.

Looking forward to create a new NuGet release.

TwitchPlaysBF2 commented:
Tests are failing because they need the following attribute
```csharp
[Explicit("Only run this test on a real Windows machine, for end-to-end testing.")]
``` 

Since the test-runner doesn't have the game installed.

They are meant to only explicitly be ran - when testing on a windows machine

TwitchPlaysBF2 commented on src/GameDataReader.Tests/Battlefield1942/GameDataReadersTests.cs:
```diff
@@ -37,4 +40,35 @@ public void GameDataReaders_GetAccessor_IsNotSingleton()
+
+        Assert.IsTrue(File.Exists(filePath));
+    }
+
```
Add this, so the test runner on GitHub doesn't fail:
```csharp
[Explicit("Only run this test on a real Windows machine, for end-to-end testing.")]
```

TwitchPlaysBF2 commented on src/GameDataReader.Tests/BattlefieldVietnam/GameDataReadersTests.cs:
```diff
@@ -37,4 +39,35 @@ public void GameDataReaders_GetAccessor_IsNotSingleton()
         bfVietnamDataReader1.GetHashCode().Should()
             .NotBe(bfVietnamDataReader2.GetHashCode());
     }
+
```
Add this, so the test runner on GitHub doesn't fail:
```csharp
[Explicit("Only run this test on a real Windows machine, for end-to-end testing.")]
```

TwitchPlaysBF2 commented on src/GameDataReader.Tests/BattlefieldVietnam/GameDataReadersTests.cs:
```diff
@@ -37,4 +39,35 @@ public void GameDataReaders_GetAccessor_IsNotSingleton()
+        Assert.IsTrue(File.Exists(filePath));
+    }
+
+    [Test]
```
Add this, so the test runner on GitHub doesn't fail:
```csharp
[Explicit("Only run this test on a real Windows machine, for end-to-end testing.")]
```

AnomalousNicole replied:
Hey @TwitchPlaysBF2

Thank you for the kind words regarding my issue summary and the fixes that I've worked on so far.
I was glad to run across Battlefield Rich Presence, while I was searching for Discord bots that I could add to the MoonGamers Battlefield 1942 Discord server.  When I discovered that Battlefield Rich Presence didn't seem to work with Battlefield 1942, I decided that I wanted to figure out why and see if I could get it working.  I am still kind of a newbie when it comes to programming in C#. :)  Prior to my current job, I had been programming in Borland Delphi 5 (Object Pascal) for 10 years on a legacy Windows XP era program.  Visual Studio and .NET have been a breath of fresh air compared to the obsolete Borland Delphi 5 IDE I had to use.


**Regarding the kernel32 DLL import/GetMainModuleFileName change:**
Now that I take a second look at the code and at what I was trying to accomplish, I agree that the "Process.GetProcessesByName" code above will work correctly for getting the path of the game process executable.

When I was looking for code to get the path of the game process executable, I was trying to take into account that the games could be running under 32bit or 64bit operating systems.
I realize now, that BF1942 and BFV are both 32bit games and the solution below is not necessary and the code can be simplified.
https://stackoverflow.com/questions/5497064/how-to-get-the-full-path-of-running-process/48319879#48319879

TwitchPlaysBF2 replied:
Delphi wow 😀 well I hope C# is not the Delphi of the future 🤣

Regarding the simplification - alright, sounds good!

TwitchPlaysBF2 reviewed (changes requested):
Tests are failing because the test-runner doesn't have the game installed.
They are meant to only be ran explicitly - while testing on a windows machine

AnomalousNicole committed:
fix: Made the fixes below based on feedback from the initial PR commit

Added the Explicit attribute to the BF1942 and BFV unit tests that read the con files while the game is running
Moved the GetProcessPath and GetConFilePath calls to the GetFilePath methods for the Global and Profile V1 Refractor config files
Refactored UtilityMethodsRefractorV1ConfigFile to RefractorV1PathResolver

AnomalousNicole committed:
fix: Refactored the RefractorV1PathResolver methods to try and read the con file path using the Registry, Game process, or AppData paths

AnomalousNicole committed:
fix: Simplified the RefractorV1PathResolver methods below and removed the 32bit Operating System code

GetPathFromRegistry
GetPathFromGameProcess
GetPathFromAppData

AnomalousNicole committed:
fix: Fixed con file path lookup bugs in the methods below used to find the bf1942 and bfv con files

GetPathFromRegistry
GetPathFromGameProcess
GetPathFromAppData

AnomalousNicole commented:
@cetteup @TwitchPlaysBF2 

**I have pushed some commits that I believe address the items that were noted regarding the original PR commit.
GameDataReader will now try to read the con files from the locations listed in the order below.**

**Registry:** If the game was installed properly, the con file path will be read from the game's GAMEDIR value in the Registry.
The Registry lookup will check if the con file exists and if it does exist then the lookup method will return that path to the method that invoked the lookup.

**Game Process:** If the game was not properly installed, the con file path will be read using the game process path.
The Game Process lookup will check if the con file exists and if it does exist then the lookup method will return that path to the method that invoked the lookup.

**AppData:** If the con files could not be found using the Registry or game process path, then the con file path will be read using the AppData path.  The AppData lookup will check if the con file exists and if it does exist then the lookup method will return that path to the method that invoked the lookup.

**I have tested all the scenarios above using BF1942 and BFV and it appears that each lookup method is working correctly.  Please let me know if you have any questions regarding the changes.**

TwitchPlaysBF2 commented:
Thanks for your efforts! I have been offline since I was sick, but the changes look mergeable to me 🚀

AnomalousNicole commented:
@TwitchPlaysBF2 You're welcome. :)  I enjoyed working on the changes and coming up with a solution.

TwitchPlaysBF2 commented:
I've created a new release 0.0.4 on NuGet
